enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/HandshakeClientGate.java

```diff
@@ -20,59 +20,69 @@
  */
-public class HandshakeGate implements ChannelInterceptor
+public class HandshakeClientGate extends ChannelDuplexHandler
 {
-    public static final String HANDSHAKE_GATE = "HandshakeGate";
-
     private final CompletableFuture<Void> handshakePromise = new CompletableFuture<>();
+    private final Channel channel;
+    private final Log log;
 
-    HandshakeGate( Channel channel, Log log )
+    public HandshakeClientGate( Channel channel, Log log )
+    {
+        this.channel = channel;
+        this.log = log;
+    }
+
+    @Override
+    public void userEventTriggered( ChannelHandlerContext ctx, Object evt ) throws Exception
     {
-        log.info( "Handshake gate added" );
-        channel.pipeline().addFirst( HANDSHAKE_GATE, new ChannelInboundHandlerAdapter()
+        if ( HandshakeFinishedEvent.getSuccess().equals( evt ) )
         {
-            @Override
-            public void userEventTriggered( ChannelHandlerContext ctx, Object evt ) throws Exception
-            {
-                if ( HandshakeFinishedEvent.getSuccess().equals( evt ) )
-                {
-                    log.info( "Handshake gate success" );
-                    handshakePromise.complete( null );
-                }
-                else if ( HandshakeFinishedEvent.getFailure().equals( evt ) )
-                {
-                    log.warn( "Handshake gate failed" );
-                    handshakePromise.completeExceptionally( new ClientHandshakeException( "Handshake failed" ) );
-                    channel.close();
-                }
-                else
-                {
-                    super.userEventTriggered( ctx, evt );
-                }
-            }
-        } );
+            log.info( "Handshake gate success" );
+            handshakePromise.complete( null );
+        }
+        else if ( HandshakeFinishedEvent.getFailure().equals( evt ) )
+        {
+            log.warn( "Handshake gate failed" );
+            handshakePromise.completeExceptionally( new ClientHandshakeException( "Handshake failed" ) );
+            channel.close();
+        }
+        else
+        {
+            super.userEventTriggered( ctx, evt );
+        }
     }
 
     @Override
-    public void write( BiFunction<Channel,Object,Future<Void>> writer, Channel channel, Object msg, CompletableFuture<Void> promise )
+    public void write( ChannelHandlerContext ctx, Object msg, ChannelPromise promise ) throws Exception
     {
-        handshakePromise.whenComplete( ( ignored, failure ) ->
-                writer.apply( channel, msg ).addListener( x -> promise.complete( null ) ) );
+        if ( handshakePromise.isDone() || msg instanceof ServerMessage )
+        {
+            ctx.write( msg, promise );
+        }
+        else
+        {
+            handshakePromise.whenComplete( ( ignored, failure ) -> {
+                if ( failure == null )
+                {
+                    channel.writeAndFlush( msg, promise );
```

Comment:
Should this be calling a method on channel or on ctx? Should we be flushing from within a write method? Should we also override the flush method on ChannelDuplexHandler?

Reply:
I actually want to remake this a bit (later) so that the messages are queued up on an actual queue and then the entire thing has a bound and gets drained completely with writes and a single flush.

I will take another look at channel vs. ctx and clarify with a comment which is correct and why considering the pipeline directions and the position of the gate. It looks potentially dubious to me as it is.

Reply:
I figured out what the issue here is. The "fixed" handler concept in the pipeline builder has really crappy semantics, since you really have no idea about which position they get on a reinstall. Right now when installing the handshaking it is added to end the which is the correct thing to gate the outbound (last->first), but after a reinstall it will have moved relative position to the beginning and thus channel is needed instead of context to get the queued messages through the pipeline. Of course the issue is with the design of the "fixed" handler concept. I will try to come up with something at least a tad bit more robust, perhaps tail/head fixed handlers. Ideas welcome.